### ISSUE TITLE
Missing or suspicious image alt text (severity 10)

### ISSUE THREAD

**Issue by jvine** [sul-dlss/exhibits#1585]
Image alt text is the most severe and most common a11y issue in Exhibits. There are a few different categories of image failures, in order of priority:

## 1. Image has no alt text, and is a link (= no link text)
Examples:
- [Exhibit item widget](https://exhibits-stage.stanford.edu/herbert-matter)
- [Search results - masonry](https://exhibits-stage.stanford.edu/lanciani/catalog?q=&search_field=search&utf8=%E2%9C%93&view=masonry)

Suggested solution: 
- [x] Automatically add the image title as the alt text. 

### 1 b. ... and also has a visible linked caption
Examples:
- [Browse category](https://exhibits-stage.stanford.edu/herbert-matter/browse/alexander-calder)
- [Search results - list](https://exhibits.stanford.edu/lanciani/catalog?q=&search_field=search&utf8=%E2%9C%93&view=list)
- [Search results - gallery](https://exhibits.stanford.edu/lanciani/catalog?q=&search_field=search&utf8=%E2%9C%93&view=gallery)

This is more complicated. There are two links in the HTML: 
1. the image itself (no alt; empty link text)
2. the caption or title below/next to the image.

Two links to the same item is already an issue; adding alt text to the images will arguably make the experience worse.

My suggestion here is:
- [x] Give the image an empty `alt=""` attribute (though it wouldn't be harmful to use the title)
- [x] Add `aria-hidden="true"` to the `<a>` around the image
- [x] Remove that `<a>` from the tab order with tabindex="-"`

Outcome: Only the linked caption is presented to the screen reader or keyboard user. Mouse or touch user can still click the image.

## 2. Image has no alt text, is not a (separate) link
Example:
- [Contact image](https://exhibits-stage.stanford.edu/herbert-matter/about/herbert-matter) (image not linked)
- [Browse categories](https://exhibits-stage.stanford.edu/herbert-matter/browse) (image is inside the same `<a>` as the text)

I think it's reasonable to treat these images as decorative.

Suggested solution:
- [x] "hide" the image from SR by adding `alt=""`

## 3. Image has "suspicious" alt text
Example:
- [Uploaded item row](https://exhibits-stage.stanford.edu/lanciani)

Images uploaded into this widget seem to end up with their filename as the alt attribute. This is never going to be good alt text:  lanciani275.jpg, Untitled.png, etc.

The correct solution here would be to require the user to add alt text on image upload. In the immediate term... it might be preferable to treat the images as decorative. If the user creates a caption, it will be visible/read. If they don't, then the image is effectively not present. Open to discussion on this.

Possible solution:
- [x] replace `alt="filename.jpg"` with `alt=""`

**Comment by ggeisler** [comment]
For item 3, the uploaded item row: the upload item form does include a title field. It's not currently a required field, but it does seem reasonable to require it (right?), which would guarantee that we have a title to use for the alt text. That would be an easy fix (for new uploaded items, anyway), but is only half of the problem.

We also have a bulk upload item option that takes a spreadsheet, and of course we can't refuse to upload the spreadsheet if the user omits all the title values in the spreadsheet (because I don't think we can check the spreadsheet before it is uploaded). So that might be why we don't already require a title. We could probably refuse to add the uploaded items from the spreadsheet if they don't have a title (I'm guessing that's technically feasible). But that would take some development work.

**Comment by jvine** [comment]
I only saw a Caption field in the widget, which ... I mean, I think Exhibits needs to distinguish between visible captions and alt text, which might reasonably be different things (which you also said in your edit, hah). Empty alt text is a safe fallback as long as the image isn't going to be a standalone link.

**Comment by ggeisler** [comment]
Ah, wait, sorry, I was thinking about uploaded items that the curator imports into the exhibit (via the dashboard and a different form), after which they are treated as regular items -- so the Upload item row widget isn't relevant. So ignore my comment; as you say we should just to the `alt=" "` solution there.
